docs/src/tutorial.rst

```diff
@@ -0,0 +1,496 @@
+
+   This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+   either express or implied. See the License for the specific language governing permissions and
+   limitations under the License.
+
+.. _tutorial:
+
+##############
+|cdk| Tutorial
+##############
+
+This topic steps you through creating the resources for a simple widget dispensing
+service using the |cdk|.
+
+.. _overview:
+
+Overview
+========
+
+This tutorial contains the following steps.
+
+1. Create a |cdk| app
+
+2. Create a |LAMlong| function that gets a list of widgets with: GET /
+
+3. Create the service that calls the |LAM| function
+
+4. Add the service to the |cdk| app
+
+5. Test the app
+
+6. Add |LAM| functions to:
+
+    * create an widget based with: POST /{name}
+    * get an widget by name with: GET /{name}
+    * delete an widget by name with: DELETE /{name}
+
+.. _create_app:
+
+Step 1: Create a |cdk| App
+==========================
+
+Let's create the TypeScript app **MyWidgetService** in in the current folder.
+
+.. code-block:: sh
+
+    mkdir MyWidgetService
+    cd MyWidgetService
+    cdk init --language typescript
+
+This creates *my_widget_service.ts* in the *bin* directory.
+We don't need most of this code,
+so for now change it to the following:
+
+.. code-block:: ts
+
+    #!/usr/bin/env node
+    import cdk = require('@aws-cdk/cdk');
+
+    class MyWidgetServiceStack extends cdk.Stack {
+      constructor(parent: cdk.App, name: string, props?: cdk.StackProps) {
+        super(parent, name, props);
+
+
+      }
+    }
+
+    // Create a new CDK app
+    const app = new cdk.App();
+
+    // Add your stack to it
+    new MyWidgetServiceStack(app, 'MyWidgetServiceStack');
+
+    app.run();
+
+Save it and make sure it builds and creates an empty stack.
+
+.. code-block:: sh
+
+    npm run build
+    cdk synth
+
+You should see a stack like the following,
+where CDK-VERSION is the version of the CDK.
+
+.. code-block:: sh
+
+    Resources:
+      CDKMetadata:
+        Type: 'AWS::CDK::Metadata'
+        Properties:
+          Modules: >-
+            @aws-cdk/cdk=CDK-VERSION,@aws-cdk/cx-api=CDK-VERSION,my_widget_service=0.1.0
+
+.. _create_lambda_functions:
+
+Step 2: Create a |LAM| Function to List all Widgets
+===================================================
+
+The next step is to create a |LAM| function to list all of the widgets in our
+|S3| bucket.
+
+Create the directory *resource* at the same level as the *bin* directory.
+
+.. code-block:: sh
+
+    mkdir resources
+
+Create the following Javascript file, *widgets.js*,
+in the *resources* directory.
+
+.. code-block:: js
+
+    const AWS = require('aws-sdk');
+    const S3 = new AWS.S3();
+
+    const bucketName = process.env.BUCKET;
+
+    // For now we only handle GET /:
+    exports.main = function(event, context, callback) {
+      if (event.httpMethod == "GET") {
+        if (event.path.length == 1) {
+          S3.listObjectsV2({ Bucket: bucketName })
```

Comment:
Use NodeJS8 and you can rewrite this to be `async`, which is way more readable:

```ts
exports.main = async function(event, context) {
  try {
    if (event.httpMethod == "GET") {
      if (event.path.length == 1) {
        const data = await S3.listObjectsV2({ Bucket: bucketName }).promise();
        var body = { widgets: data.Contents.map(function(e) { return e.Key }) };
        return {
          statusCode: 200,
          headers: {},
          body: JSON.stringify(body)
        };
      }
    }
  } catch(error) {
    var body = error.stack || JSON.stringify(error, null, 2);
    var response = makeError(body);
    return response;
  }
}
```

Reply:
By the way where does `makeError` come from?

Reply:
Okay, I no longer use helper functions.

Reply:
That's not the point. I like helper functions. I just didn't see it defined. Or did I miss it?